### ISSUE TITLE
perf: FP64 Conv2DBackward at ImageNet shapes — ResNet50 / VGG16 train step 8× too slow on CPU

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#415]
## Summary

AiDotNet's ResNet50<double> and VGG16<double> training step is ~8× too slow to fit the AiDotNet-side CI test budgets at ImageNet input scale (224×224×3, 1000-class output). The dominant cost is in this repo: `Conv2DBackwardInput` + `Conv2DBackwardKernel` at FP64 across the ResNet/VGG conv stack.

## Concrete measurement

On master (`4eb7cd74f`, locally on a Ryzen-class CPU, .NET 10):

- `AiDotNet.Tests.ModelFamilyTests.NeuralNetworks.ResNetNetworkTests.Training_ShouldChangeParameters` (1 warmup forward + **10 Train calls**): **39 s** → ~3.9 s / step
- `MoreData_ShouldNotDegrade` (**250 Train calls**): hits the 120 s `[Fact(Timeout = 120000)]` budget → fails
- `LossStrictlyDecreasesOnMemorizationTask` (longer iter count): hits 180 s budget → fails
- Same shape profile in VGG16 — `Training_ShouldReduceLoss` (30 iters) and `LossStrictlyDecreasesOnMemorizationTask` both timeout

To make `MoreData_ShouldNotDegrade` fit its 120 s budget at 250 iters: need ≤ 480 ms / step → **~8× speedup** vs the current 3.9 s.

## Shape catalog (the hot path)

ResNet50 + VGG16 at 224×224 input drive the following per-layer `Conv2D` + `Conv2DBackwardInput` + `Conv2DBackwardKernel` shapes through `CpuEngine` (per the issue #1390 substep profile pattern; the equivalent shape catalog for these networks can be dumped via the `ShapeInstrumenter` landing in PR #412):

| Layer | Spatial | InChannels | OutChannels | Kernel | Stride |
|---|---|---:|---:|---|---:|
| ResNet50 conv1 | 224×224 | 3 | 64 | 7×7 | 2 |
| ResNet50 res2 (×3) | 56×56 | 64 / 256 | 64 / 256 | 1×1 + 3×3 + 1×1 | 1 |
| ResNet50 res3 (×4) | 28×28 | 128 / 512 | 128 / 512 | 1×1 + 3×3 + 1×1 | 1-2 |
| ResNet50 res4 (×6) | 14×14 | 256 / 1024 | 256 / 1024 | 1×1 + 3×3 + 1×1 | 1-2 |
| ResNet50 res5 (×3) | 7×7 | 512 / 2048 | 512 / 2048 | 1×1 + 3×3 + 1×1 | 1-2 |
| VGG16 conv block 1 | 224×224 | 3 / 64 | 64 | 3×3 | 1 |
| VGG16 conv block 2 | 112×112 | 64 / 128 | 128 | 3×3 | 1 |
| VGG16 conv block 3 | 56×56 | 128 / 256 | 256 | 3×3 | 1 |
| VGG16 conv block 4 | 28×28 | 256 / 512 | 512 | 3×3 | 1 |
| VGG16 conv block 5 | 14×14 | 512 | 512 | 3×3 | 1 |

All at batch=1 (`InputShape = [1, 3, 224, 224]` in the test fixtures), all at FP64. The forward path appears to hit the existing `Conv2DWithIm2ColDouble` fast path; the backward path is the dominant cost.

## What's likely contributing

1. **Im2col + DGEMM blow-up at small batch**: Backward needs a transposed im2col of `gradOutput` (the `Conv2DBackwardKernel` gradient-of-weight path) and re-runs the GEMM at shape `(outC, inC × kH × kW)` × `(inC × kH × kW, outH × outW)`. Im2col alloc is O(outH × outW × inC × kH × kW) doubles — for early ResNet layers at 56×56 with inC=256, kH=kW=3, the buffer is ~ 8 MB per slice per Conv2D backward call. Allocator pressure + L2 churn.
2. **No direct-conv fast path for 3×3 stride-1 backward**: The existing `Conv3x3Stride1Double` direct kernel (used in PR #410) handles forward only at that shape. Backward still goes through im2col + DGEMM.
3. **Backward-of-weight reduction order**: `Conv2DBackwardKernel` sums over the batch × output-spatial axes. At batch=1 the sum is just the per-position contribution but the allocation pattern still walks all positions.
4. **FP64 SIMD lane count**: AVX2 = 4 lanes for double vs 8 for float. Tensors-side Conv2D backward may have only an FP32 fast path tuned, with FP64 falling back to scalar or 4-wide.

## Expected scope

Mirror the 3-phase plan from #403 (DCGAN perf) but applied to ResNet50/VGG16 shapes:

- **Phase A — Profile** (using `ShapeInstrumenter` + per-substep stopwatch landing in PR #412):
  shape-catalog dump, per-conv-shape stopwatch on `Conv2D` / `Conv2DBackwardInput` / `Conv2DBackwardKernel` for one ResNet50/VGG16 step, allocation profile under `GC.GetAllocatedBytesForCurrentThread`. Output ranks the bottlenecks.
- **Phase B — `Conv2DBackward` direct-kernel gate** (most likely highest ROI):
  add a direct-conv backward path (no im2col) for 3×3 stride=1 padding=1 — the dominant VGG16 shape — and possibly 1×1 (the dominant ResNet50 bottleneck-block shape). Mirror the FP64 gate work landed in PR #410 for forward.
- **Phase C — Im2col buffer reuse**:
  `[ThreadStatic]` im2col scratch buffer sized to the largest seen shape eliminates the per-call rent/return.
- **Phase D — FP64 AVX2 microkernel quality**:
  verify `Avx2Fp64_4x8` / `Avx2Streaming` fire on the backward GEMM shapes; close any FP64 scalar fallbacks.

## Affected callers (AiDotNet repo)

- `ResNetNetworkTests.MoreData_ShouldNotDegrade` (current visible victim, **#1306**)
- `ResNetNetworkTests.Training_ShouldReduceLoss` (**#1306**)
- `VGGNetworkTests.LossStrictlyDecreasesOnMemorizationTask` (**#1306**)
- `VGGNetworkTests.TrainingError_ShouldNotExceedTestError` (**#1306**)
- `VGGNetworkTests.Training_ShouldReduceLoss` (**#1306**)
- `DenseNetNetworkTests.MoreData_ShouldNotDegrade` (**#1314** cluster 6)
- All other Cluster-6 long-training timeouts that route through Conv2D — see ooples/AiDotNet#1314 for the full list

The AiDotNet-side task #95 ("Profile + fix ResNet50 Train perf bug, target: 30 iters in 120s") tracks the same shape stack from the consumer side.

## NOT this issue

- `Conv2DBackward` at small (non-ImageNet) shapes — already covered by other perf paths (1×1 bypass, NCHWc fast path).
- DCGAN-style ConvTranspose2D — separate cost class, tracked under #358 / #403 / #1390.

## Reproducer

```bash
cd <AiDotNet repo>
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj -c Release -f net10.0 \
  --filter "FullyQualifiedName=AiDotNet.Tests.ModelFamilyTests.NeuralNetworks.ResNetNetworkTests.Training_ShouldChangeParameters"
# ~39 s for 10 Train calls = 3.9 s/step today; target ~0.48 s/step for 120s/250-iter budget.
```

Once the `ShapeInstrumenter` from PR #412 merges, the per-shape catalog dump becomes a one-liner: wrap the test's training loop in `using var probe = new ShapeInstrumenter()` and `probe.PrintCatalog()` after.
